dali/pipeline/graph/op_graph_verifier.cc

```diff
@@ -0,0 +1,149 @@
+constexpr OpType parent_constraints<OpType::MIXED>::allowed_input_ops[];
+constexpr OpType parent_constraints<OpType::GPU>::allowed_input_ops[];
+
+namespace {
+// helper to convert static information to runtime information
+template <OpType op_type>
+std::set<OpType> GetParentConstraints() {
+  return std::set<OpType>{std::begin(parent_constraints<op_type>::allowed_parents),
+                          std::end(parent_constraints<op_type>::allowed_parents)};
+}
+}  // namespace
+
+std::vector<std::set<OpType>> ParentOpTypeConstraints() {
+  std::vector<std::set<OpType>> allowed_parents;
+  allowed_parents.resize(static_cast<int>(OpType::COUNT));
+  allowed_parents[static_cast<int>(OpType::GPU)] = GetParentConstraints<OpType::GPU>();
+  allowed_parents[static_cast<int>(OpType::CPU)] = GetParentConstraints<OpType::CPU>();
+  allowed_parents[static_cast<int>(OpType::MIXED)] = GetParentConstraints<OpType::MIXED>();
+  allowed_parents[static_cast<int>(OpType::SUPPORT)] = GetParentConstraints<OpType::SUPPORT>();
+  return allowed_parents;
+}
+
+std::vector<int> ArgumentInputConstraints() {
+  std::vector<int> allows_argument_input;
+  allows_argument_input.resize(static_cast<int>(OpType::COUNT));
+  allows_argument_input[static_cast<int>(OpType::GPU)] =
+      parent_constraints<OpType::GPU>::supports_argument_inputs;
+  allows_argument_input[static_cast<int>(OpType::CPU)] =
+      parent_constraints<OpType::CPU>::supports_argument_inputs;
+  allows_argument_input[static_cast<int>(OpType::MIXED)] =
+      parent_constraints<OpType::MIXED>::supports_argument_inputs;
+  allows_argument_input[static_cast<int>(OpType::SUPPORT)] =
+      parent_constraints<OpType::SUPPORT>::supports_argument_inputs;
+  return allows_argument_input;
+}
+
+/**
+ * @brief Check if parent nodes have compatible OpType
+ */
+void CheckParentConstraints(const OpGraph& op_graph, const OpNode& op) {
+  static const auto allowed_parent_type = ParentOpTypeConstraints();
+  for (auto parent_id : op.parents) {
+    const auto& parent = op_graph.Node(parent_id);
+    const auto& allowed_parents = allowed_parent_type[static_cast<int>(op.op_type)];
+    DALI_ENFORCE(allowed_parents.find(parent.op_type) != allowed_parents.end(),
+                 "Op " + op.instance_name + " of type " + to_string(op.op_type) +
+                     " has parent Op " + parent.instance_name +
+                     " of incompatible type: " + to_string(parent.op_type) + ". Expected one of: " +
+                     concatenate_alternatives(allowed_parent_type[static_cast<int>(op.op_type)]) +
+                     ".");
+  }
+}
+
+/**
+ * @brief Check support for Argument Inputs and if Argument Inputs are produed by Support Ops
+ */
+void CheckArgumentInputConstraints(const OpGraph& op_graph, const OpNode& op) {
+  static const auto allows_argument_input = ArgumentInputConstraints();
+  bool arg_in_allowed = allows_argument_input[static_cast<int>(op.op_type)];
+  if (!arg_in_allowed) {
```

Comment:
make sure that arg inputs are allowed for mixed ops as well

Reply:
I will do so in a followup PR, do not want to introduce changes that can break something unless it is a bug.